var/spack/repos/builtin/packages/mesa/package.py

```diff
@@ -19,6 +19,10 @@ class Mesa(MesonPackage):
     url = "https://archive.mesa3d.org/mesa-20.2.1.tar.xz"
 
     version('master', tag='master')
+    # Note: If v22.x or greater is added please leave 21.3.7 as preferred.  The swr
+    # multithreaded cpu driver was dropped 22.x and is currently necessary to get
+    # reasonable rendering performance on HPC OpenGL workloads.
```

Comment:
Could you please include some URL with more references?
Like:
Is there a bug open somewhere upstream to resolve the problem?

So that is easy to check, whether the `preferred` is still needed?

Reply:
I suppose I can reference the MR that removed in in upstream mesa that contains some of this discussion: https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/11264

It's not a bug, just a current design limitation of llvmpipe.  For software based rendering with mesa there are three available drivers: softpipe, llvmpipe, and swr.  softpipe is always available and is essentially the single threaded reference implementation for cpu rendering.  It's very very slow and not particularly useful for anything other than debugging.  The llvmpipe driver was intended to target OpenGL desktop compositing and light OpenGL game workloads which both mostly rely on pixel shaders, making little use of vertex shaders.  The shader processing in llvmpipe's implementation reflects this as pixel shaders are processed in parallel which vertex shaders are processed single threaded.  While this works okay for the intended desktop compositing and game workloads, it's terrible for scientific computing and visualization workloads which are mostly reliant on vertex shaders.  To address this Intel implemented and contributed OpenSWR to the mesa project in v12 as a fully multi threaded and vectorized cpu render for all shader types.  This has been the driver of choice for cpu-based rendering in HPC for the past 6y as it's an order of magnitude more performant.  Eventually Intel decided to stop maintaining it and it was just recently remove from the main mesa source tree.